sinks/cortex/cortex.go

```diff
@@ -207,66 +208,32 @@ func (s *CortexMetricSink) Flush(ctx context.Context, metrics []samplers.InterMe
 		return sinks.MetricFlushResult{}, nil
 	}
 
-	if s.batchWriteSize == 0 || len(metrics) <= s.batchWriteSize {
-		err := s.writeMetrics(ctx, metrics)
-		if err == nil {
-			flushedMetrics = len(metrics)
-		} else {
-			s.logger.Error(err)
-			droppedMetrics = len(metrics)
-		}
-
-		return sinks.MetricFlushResult{MetricsFlushed: flushedMetrics, MetricsDropped: droppedMetrics}, err
+	var allErrs *multierror.Error
+	batchSize := s.batchWriteSize
+	if s.batchWriteSize == 0 {
+		batchSize = len(metrics)
 	}
-
-	doIfNotDone := func(fn func() error) error {
+batching:
+	for i := 0; i < len(metrics); i += batchSize {
+		end := i + batchSize
+		if end > len(metrics) {
+			end = len(metrics)
+		}
+		batch := metrics[i:end]
 		select {
 		case <-ctx.Done():
-			return errors.New("context finished before completing metrics flush")
+			droppedMetrics += len(metrics[i:])
```

Comment:
_probably_ not important, but since we are looking at performance improvements anyway...
I think `metrics[i:]` allocates a single pointer. We could avoid the allocation by...
```suggestion
			droppedMetrics += len(metrics)-i
```